solver/include/solver/solver.hpp

```diff
@@ -4,15 +4,308 @@
+    {
+        using difference_type   = value_t;
+        using value_type        = current_solution<value_t, state_t>;
+        using pointer           = value_type*;
+        using reference         = value_type&;
+        using const_pointer     = value_type const*;
+        using const_reference   = value_type const&;
+        using iterator_category = std::output_iterator_tag;
+
+        value_type sol;
+        method_t meth;
+        problem_t& pb;
+        ponio::time_span<value_t> t_span;
+        typename ponio::time_span<value_t>::iterator it_next_time;
+        value_t final_time; // should be a ponio::time_span;
+        static constexpr value_t sentinel = std::numeric_limits<value_t>::max();
+
+        // time_iterator ( problem_t & pb_, method_t meth_, state_t const& u0, ponio::time_span<value_t> && times, value_t dt )
+        // : sol(times.front(), u0, dt)
+        // , meth(meth_)
+        // , pb(pb_)
+        // , t_span(times)
+        // , it_next_time(++times.begin())
+        // {
+        // }
+
+        time_iterator( problem_t& pb_, method_t meth_, state_t const& u0, ponio::time_span<value_t> const& t_span_, value_t dt )
+            : sol( ( t_span_.front() == t_span_.back() ) ? sentinel : t_span_.front(), u0, dt )
+            , meth( meth_ )
+            , pb( pb_ )
+            , t_span( t_span_ )
+            , it_next_time( std::begin( t_span ) )
+            , final_time( t_span.back() )
+        {
+        }
+
+        time_iterator( time_iterator const& rhs )
+            : sol( rhs.sol )
+            , meth( rhs.meth )
+            , pb( rhs.pb )
+            , t_span( rhs.t_span )
+            , it_next_time( std::begin( t_span ) + ( rhs.it_next_time - std::begin( rhs.t_span ) ) )
+            , final_time( rhs.final_time )
+        {
+        }
+
+        void
+        increment()
+        {
+            std::tie( sol.time, sol.state, sol.time_step ) = meth( pb, sol.time, sol.state, sol.time_step );
+        }
+
+        difference_type
+        next_time() const
+        {
+            return sol.time + sol.time_step;
+        }
+
+        time_iterator&
+        operator++()
+        {
+            // if ( next_time() > *it_next_time )
+            // {
+            //   sol.time_step = final_time - sol.time;
+            //   ++it_next_time;
+            // }
+            if ( sol.time == final_time )
+            {
+                sol.time = sentinel;
+            }
+
+            if ( sol.time == sentinel )
+            {
+                return *this;
+            }
+
+            auto save_dt   = sol.time_step;
+            bool change_dt = false;
+            if ( next_time() > *it_next_time )
+            {
+                sol.time_step = *it_next_time - sol.time;
+                ++it_next_time;
+                change_dt = true;
+            }
+
+            increment();
+            if ( change_dt )
+            {
+                sol.time_step = save_dt;
+            }
```

Comment:
Why do you do this since it's already done in the previous if ?

Reply:
If the current method is a constant time step method, you change the dt to a value to go to the next time, but after needing to reset the time step to the previous time step, here I change the time step before returning the value (it's a bug) I fix it with a `std::optional`.